src/safeds_runner/server/memoization_map.py

```diff
@@ -0,0 +1,176 @@
+            f"Last access: {self.last_access}, computation time: {self.computation_time}, lookup time:"
+            f" {self.lookup_time}, memory size: {self.memory_size}"
+        )
+
+
+class MemoizationMap:
+    """
+    The memoization map handles memoized function calls.
+
+    This contains looking up stored values, computing new values if needed and calculating and updating statistics.
+    """
+
+    def __init__(
+        self,
+        map_values: dict[tuple[str, tuple[Any], tuple[Any]], Any],
+        map_stats: dict[tuple[str, tuple[Any], tuple[Any]], MemoizationStats],
+    ):
+        """
+        Create a new memoization map using a value store dictionary and a stats dictionary.
+
+        Parameters
+        ----------
+        map_values
+            Value store dictionary
+        map_stats
+            Stats dictionary
+        """
+        self.map_values: dict[tuple[str, tuple[Any], tuple[Any]], Any] = map_values
+        self.map_stats: dict[tuple[str, tuple[Any], tuple[Any]], MemoizationStats] = map_stats
+
+    def memoized_function_call(
+        self,
+        function_name: str,
+        function_callable: Callable,
+        parameters: list[Any],
+        hidden_parameters: list[Any],
+    ) -> Any:
+        """
+        Handle a memoized function call.
+
+        Looks up the stored value, determined by function name, parameters and hidden parameters and returns it if found.
+        If no value is found, computes the value using the provided callable and stores it in the map.
+        Every call to this function will update the memoization stats.
+
+        Parameters
+        ----------
+        function_name
+            Fully qualified function name
+        function_callable
+            Function that is called and memoized if the result was not found in the memoization map
+        parameters
+            List of parameters passed to the function
+        hidden_parameters
+            List of hidden parameters for the function. This is used for memoizing some impure functions.
+
+        Returns
+        -------
+        The result of the specified function, if any exists
+        """
+        key = (function_name, _convert_list_to_tuple(parameters), _convert_list_to_tuple(hidden_parameters))
+        time_compare_start = time.perf_counter_ns()
+        try:
+            potential_value = self.map_values[key]
+        except KeyError:
+            pass
+        else:
+            time_compare_end = time.perf_counter_ns()
+            # Use time_ns for absolute time points, as perf_counter_ns does not guarantee any fixed reference-point
+            time_last_access = time.time_ns()
+            time_compare = time_compare_end - time_compare_start
+            old_memoization_stats = self.map_stats[key]
+            memoization_stats = MemoizationStats(
+                time_last_access,
+                old_memoization_stats.computation_time,
+                time_compare,
```

Comment:
Instead of discarding the old value completely, consider using something like an [exponential moving average](https://en.wikipedia.org/wiki/Exponential_smoothing) or even a list of all access times. The latter means we don't lose information at the cost of higher memory and higher cost to compute a single value used as a decider later.